### PR TITLE
abstract Quote class from Exchange

### PR DESCRIPTION
**Pull request by wangwenxi-handsome**
## How Has This Been Tested?
- [ ] Pass the test by running: `pytest qlib/tests/test_all_pipeline.py` under upper directory of `qlib`.
- [ ] If you are adding a new feature, test on your own test scripts.

## Screenshots of Test Results (if appropriate):
1. Pipeline test:
2. Your own tests:

## Types of changes

- [ ] Fix bugs
- [x] Add new feature
- [ ] Update documentation

### CODE REVIEW

File: qlib/backtest/exchange.py [hunk 1]
```diff
@@ -140,39 +140,48 @@ def __init__(
...
             # The `factor.day.bin` file exists and all data `close` and `factor` are not `nan`
             # Use normal price
             self.trade_w_adj_price = False
 
         # update limit
-        self._update_limit()
-
-        quote_df = self.quote
+        self._update_limit(self.limit_threshold)
+
```
Comment (you-n-g) [inline qlib/backtest/exchange.py:183]: please remove extra blank

File: qlib/backtest/exchange.py [hunk 2]
```diff
@@ -615,7 +615,7 @@ def get_data(
         start_time: Union[pd.Timestamp, str],
         end_time: Union[pd.Timestamp, str],
         fields: Union[str, list] = None,
-        method: Union[str, Callable] = None,
+        method: Union[str, "Callable"] = None,
```
Comment (you-n-g) [inline qlib/backtest/exchange.py:618]: Why do you have to use string instead of `typing.Callable` ?

**Comment by you-n-g** [comment]
Please resolve the conflictions 
image

File: qlib/backtest/report.py [hunk 3]
```diff
@@ -288,66 +291,50 @@ def _update_order_trade_info(self, trade_info: list):
...
-        self.order_indicator["pa"] = 0
+        def func(deal_amount, amount):
+            return deal_amount / amount
+        self.order_indicator.transfer(func, "ffr")
 
     def update_order_indicators(self, trade_info: list):
         self._update_order_trade_info(trade_info=trade_info)
         self._update_order_fulfill_rate()
-        self._update_order_price_advantage()
+        # self._update_order_price_advantage()
```
Comment (you-n-g) [inline qlib/backtest/report.py:314]: Why do we have to comment out this line?

File: qlib/backtest/report.py [hunk 4]
```diff
@@ -288,66 +291,50 @@ def _update_order_trade_info(self, trade_info: list):
...
-        self.order_indicator["trade_cost"] = trade_cost
-        self.order_indicator["trade_dir"] = trade_dir
+        def func_apply(trade_dir):
+            return trade_dir.apply(Order.parse_dir)
+        self.order_indicator.transfer(func_apply, "trade_dir")
 
     def _update_trade_amount(self, outer_trade_decision: BaseTradeDecision):
         # NOTE: these indicator is designed for order execution, so the
         decision: List[Order] = outer_trade_decision.get_decision()
         if decision is None:
```
Comment (you-n-g) [inline qlib/backtest/report.py:334]:
```suggestion
        if len(decision) == 0:
```
fix previous bug

File: qlib/backtest/report.py [hunk 5]
```diff
@@ -479,55 +469,58 @@ def agg_order_indicators(
...
             raise ValueError(f"method {method} is not supported!")
+        return self.order_indicator.transfer(func)
 
     def _cal_trade_positive_rate(self):
-        pa_order = self.order_indicator["pa"]
-        if isinstance(pa_order, (int, float)):
-            # pa from atomic executor
-            return pa_order
-        return (pa_order > 0).astype(int).sum() / pa_order.count()
+        def func(pa):
```
Comment (you-n-g) [inline qlib/backtest/report.py:506]: (optional)  you can consider using lambda function to simplify the code

File: qlib/backtest/report.py [hunk 6]
```diff
@@ -560,3 +553,236 @@ def get_trade_indicator(self):
...
+        col : str
+            the metric name of one metric.
+        metric : Union[dict, pd.Series]
+            the metric data.
+        """
+
+        pass
+
+    def transfer(self, func: "Callable", new_col: str = None):
+        """compute new metric with existing.
```
Comment (you-n-g) [inline qlib/backtest/report.py:579]:
```suggestion
        """compute new metric with existing ones.
```

File: qlib/backtest/report.py [hunk 7]
```diff
@@ -560,3 +553,236 @@ def get_trade_indicator(self):
 
     def generate_trade_indicators_dataframe(self):
         return pd.DataFrame.from_dict(self.trade_indicator_his, orient="index")
+
+
+class BaseOrderIndicator:
+    """The data structure of order indicator.
+    """
+
+    def __init__(self):
```
Comment (you-n-g) [inline qlib/backtest/report.py:562]: If it is not part of the interface, I think it is not necessary

File: qlib/backtest/report.py [hunk 8]
```diff
@@ -560,3 +553,236 @@ def get_trade_indicator(self):
...
+        ----------
+        col : str
+            the metric name of one metric.
+        metric : Union[dict, pd.Series]
+            the metric data.
+        """
+
+        pass
+
+    def transfer(self, func: "Callable", new_col: str = None):
```
Comment (you-n-g) [inline qlib/backtest/report.py:578]: Why not `typing.Callable`?

File: qlib/backtest/report.py [hunk 9]
```diff
@@ -560,3 +553,236 @@ def get_trade_indicator(self):
...
+            the kwargs of func will be replaced with metric data by name in this function.
+            e.g.
+                def func(pa):
+                    return (pa > 0).astype(int).sum() / pa.count()
+        new_col : str, optional
+            New metric will be assigned in the data if new_col is not None, by default None.
+
+        Return 
+        ----------
+        SingleMetric
```
Comment (you-n-g) [inline qlib/backtest/report.py:594]: Only return value when new_col is None?

Comment (you-n-g) [reply qlib/backtest/report.py:594]: Please note `SingleMetric` is optional.

File: qlib/backtest/report.py [hunk 10]
```diff
@@ -560,3 +553,236 @@ def get_trade_indicator(self):
...
+
+        def __add__(self, other):
+            if isinstance(other, (int, float)):
+                return PandasOrderIndicator.SingleMetric(self.metric + other)
+            elif isinstance(other, PandasOrderIndicator.SingleMetric):
+                return PandasOrderIndicator.SingleMetric(self.metric + other.metric)
+            else:
+                return NotImplemented
+
+        def __radd__(self, other):
```
Comment (you-n-g) [inline qlib/backtest/report.py:666]: I think you can just return `other + self`
Same as other `__r*__` functions

File: qlib/backtest/report.py [hunk 11]
```diff
@@ -560,3 +553,236 @@ def get_trade_indicator(self):
...
+                return PandasOrderIndicator.SingleMetric(self.metric > other)
+            elif isinstance(other, PandasOrderIndicator.SingleMetric):
+                return PandasOrderIndicator.SingleMetric(self.metric > other.metric)
+            else:
+                return NotImplemented
+        
+        def __len__(self):
+            return len(self.metric)
+
+        def sum(self):
```
Comment (you-n-g) [inline qlib/backtest/report.py:733]: these functions are relied on by the framework.
But there is no corresponding interface.

File: qlib/backtest/report.py [hunk 12]
```diff
@@ -560,3 +553,236 @@ def get_trade_indicator(self):
...
+            self.data[new_col] = tmp_metric
+        return tmp_metric
+
+    def get_metric_series(self, metric: str):
+        if(metric in self.data):
+            return self.data[metric].metric
+        else:
+            return pd.Series()
+
+    @staticmethod
```
Comment (you-n-g) [inline qlib/backtest/report.py:778]: In the parent class,  it is `classmethod`

**Review by you-n-g — COMMENTED** [review]
Nice job.
A lot of previous details are simplified.

File: qlib/backtest/exchange.py [hunk 13]
```diff
@@ -596,3 +524,198 @@ def get_order_helper(self) -> OrderHelper:
...
+        """add limitation information to data based on limit_threshold
+        """
+        raise NotImplementedError(f"Please implement the `_update_limit` method")
+
+    def get_trade_w_adj_price(self):
+        """return whether use the trade price with adjusted weight
+        """
+        raise NotImplementedError(f"Please implement the `get_trade_w_adj_price` method")
+
+    def get_all_stock(self):
```
[the remark this answers is not in the document]
Comment (you-n-g) [reply qlib/backtest/exchange.py:602]:
```suggestion
    def get_all_stock(self) ->  Union[list, Dict.keys(), set, tuple]:
```

File: qlib/backtest/report.py (hunk 9 above)
Comment (you-n-g) [reply qlib/backtest/report.py:594]: It is not forced to be used in internal implementation.

File: qlib/backtest/report.py [hunk 14]
```diff
@@ -560,3 +553,236 @@ def get_trade_indicator(self):
...
+    def transfer(self, func: "Callable", new_col: str = None):
+        """compute new metric with existing.
+
+        Parameters
+        ----------
+        func : Callable
+            the func of computing new metric.
+            the kwargs of func will be replaced with metric data by name in this function.
+            e.g.
+                def func(pa):
```
Comment (you-n-g) [inline qlib/backtest/report.py:587]: What could `pa` be?
NOTE: 
- SingleMetric is not enforced.

File: qlib/backtest/high_performance_ds.py [hunk 15]
```diff
@@ -0,0 +1,414 @@
...
+    """
+    The data structure of the single metric.
+    The following methods are used for computing metrics in one indicator.
+    """
+
+    def __init__(self, metric: Union[dict, pd.Series]):
+        pass
+
+    def __add__(self, other: Union["BaseSingleMetric", int, float]) -> "BaseSingleMetric":
+        pass
```
Comment (you-n-g) [inline qlib/backtest/high_performance_ds.py:125]: raise  not implement error

File: qlib/backtest/high_performance_ds.py [hunk 16]
```diff
@@ -0,0 +1,414 @@
...
+    def __len__(self) -> int:
+        pass
+
+    def sum(self) -> float:
+        pass
+
+    def mean(self) -> float:
+        pass
+
+    def count(self) -> int:
```
Comment (you-n-g) [inline qlib/backtest/high_performance_ds.py:160]: Please define count (e.g. NaN is not included)

File: qlib/backtest/high_performance_ds.py [hunk 17]
```diff
@@ -0,0 +1,414 @@
...
+    @property
+    def empty(self) -> bool:
+        """If metric is empyt, return True."""
+        pass
+
+    def add(self, other: "BaseSingleMetric", fill_value: float = None) -> "BaseSingleMetric":
+        """Replace np.NaN with fill_value in two metrics and add them."""
+        pass
+
+    def apply(self, map_dict: dict) -> "BaseSingleMetric":
```
Comment (you-n-g) [inline qlib/backtest/high_performance_ds.py:178]: I think apply is more complicated than a  map_dict

File: qlib/backtest/high_performance_ds.py [hunk 18]
```diff
@@ -0,0 +1,414 @@
...
+
+
+class BaseOrderIndicator:
+    """
+    The data structure of order indicator.
+    !!!NOTE: There are two ways to organize the data structure. Please choose a better way.
+        1. one way is use BaseSingleMetric to represent each metric. For example, the data
+        structure of PandasOrderIndicator is Dict[str: PandasSingleMetric]. It uses
+        PandasSingleMetric based on pd.Series to represent each metric.
+        2. the another way doesn't BaseSingleMetric to represent each metric. The data
```
Comment (you-n-g) [inline qlib/backtest/high_performance_ds.py:190]: Grammar

File: qlib/backtest/high_performance_ds.py [hunk 19]
```diff
@@ -0,0 +1,414 @@
...
+
+        Parameters
+        ----------
+        col : str
+            the metric name of one metric.
+        metric : Union[dict, pd.Series]
+            the metric data.
+        """
+
+        pass
```
Comment (you-n-g) [inline qlib/backtest/high_performance_ds.py:205]: Please raise NotImplementError

File: qlib/backtest/high_performance_ds.py [hunk 20]
```diff
@@ -0,0 +1,414 @@
...
+
+    def __sub__(self, other):
+        if isinstance(other, (int, float)):
+            return PandasSingleMetric(self.metric - other)
+        elif isinstance(other, PandasSingleMetric):
+            return PandasSingleMetric(self.metric - other.metric)
+        else:
+            return NotImplemented
+
+    def __rsub__(self, other):
```
Comment (you-n-g) [inline qlib/backtest/high_performance_ds.py:297]: This is not necessary now

Comment (wangwenxi-handsome) [reply qlib/backtest/high_performance_ds.py:297]: def func(trade_dir, trade_price, base_price):
                sign = 1 - trade_dir * 2
                return sign * (trade_price / base_price - 1)

this function uses it.

**Pull request merged**